pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1066,65 +1084,76 @@ impl RemoteTimelineClient {
         // into stopped state, thereby dropping all off the queued *ops* which haven't become *tasks* yet.
         // The other *tasks* will come here and observe an already shut down queue and hence simply wrap up their business.
         let mut guard = self.upload_queue.lock().unwrap();
-        let mut res = None;
-        take_mut::take(&mut *guard, |queue| {
-            match queue {
-                UploadQueue::Uninitialized => {
-                    res = Some(Err(StopError::QueueUninitialized));
-                    queue
-                }
-                UploadQueue::Stopped(_) => {
-                    // nothing to do
-                    info!("another concurrent task already shut down the queue");
-                    res = Some(Ok(()));
-                    queue
-                }
-                UploadQueue::Initialized(qi) => {
-                    let UploadQueueInitialized {
-                        task_counter: _,
-                        latest_files,
-                        // XXX need to think about what it means if it's non-zero here
-                        latest_files_changes_since_metadata_upload_scheduled: _,
-                        latest_metadata,
-                        last_uploaded_consistent_lsn,
-                        num_inprogress_layer_uploads,
-                        num_inprogress_metadata_uploads,
-                        num_inprogress_deletions,
-                        inprogress_tasks,
-                        queued_operations,
-                    } = qi;
-
-                    // consistency check
-                    assert_eq!(
-                        num_inprogress_layer_uploads
-                            + num_inprogress_metadata_uploads
-                            + num_inprogress_deletions,
-                        inprogress_tasks.len()
-                    );
+        // If any of the code below panics, the queue remains in Broken state.
+        // If we're coming from Initialized state, `queued_operations` will get dropped
+        // as part of the panic, because it sits in the local variable named `owned`.
+        // Any `wait_completion` operations against those queued operations
+        // will observe an error. That's exactly what we want.
+        // We don't need to care about in-progress operations because that responsibility
+        // lies with the caller. There's no point for them to try anything funky, like,
+        // catching the panic and retrying the stop() call. We will return QueueBroken in that case.
+        let owned = std::mem::replace(&mut *guard, UploadQueue::Broken);
```

Comment:
Looking at the latest main, this used to be `Initialized` being replaced with `Stopped`, so why is all this required in the first place?

https://github.com/neondatabase/neon/blob/291b4f0d41603c8d0ef974c81ccce8a3f1eb2589/pageserver/src/tenant/remote_timeline_client.rs#L969-L974

Of its fields, it seems that the `latest_files`, `latest_metadata`, and of course `deleted_at` are new. `last_upload_consistent_lsn` was there already.

Reply:
This was changed in 11c18b05aa67ecbe3ed3db2ae92e60da86d3bd31 but I fail to see why. Was it just to use `take_mut` or something else?

Reply:
Difficult to jump in to revert that change because many things changed.

Reply:
#4156, now looking at the race.